arbnode/espresso_batcher_addr_monitor.go

```diff
@@ -95,80 +88,50 @@ func NewBatcherAddrMonitor(
 		fromParentBlock = deployAt + 1
 	}
 	return &BatcherAddrMonitor{
-		initAddresses:             initAddresses,
 		db:                        db,
 		l1Reader:                  l1Reader,
 		seqInboxAddr:              seqInboxAddr,
 		seqInboxInterface:         seqInboxInterface,
 		deployAt:                  deployAt,
 		lastProcessedParentHeight: fromParentBlock - 1,
 		step:                      step,
+		bufferWindow:              BUFFER_WINDOW,
 	}
 }
 
-func (b *BatcherAddrMonitor) AddBatchPosterSetEvents(events []BatcherAddrUpdate) error {
-	if len(events) == 0 {
-		return nil
+func (b *BatcherAddrMonitor) IsValid(ctx context.Context, batcherAddress common.Address, l1Height uint64) (bool, error) {
+	if len(b.eventUpdatesAt) == 0 {
+		b.addEventUpdates([]uint64{b.deployAt})
 	}
-	log.Info("adding batcher addr events", "events", events)
-	b.updates = append(b.updates, events...)
-	// Sort events by l1Height to ensure correct processing order.
-	// Since BatcherAddr events are infrequent, the performance impact of sorting is negligible.
-	sort.Slice(b.updates, func(i, j int) bool {
-		return b.updates[i].L1Height < b.updates[j].L1Height
-	})
-	b.lastEventL1Height = b.updates[len(b.updates)-1].L1Height
-	b.cached = false
-	return b.Store()
-}
-
-func (b *BatcherAddrMonitor) GetValidAddresses(targetL1Height uint64) []common.Address {
-	if targetL1Height > b.lastProcessedL1Height {
-		// If the target L1 height is greater than the latest known L1 height,
-		// return an empty slice. The caller should wait until the monitor has
-		// observed at least this L1 height before calling this function.
-		return []common.Address{}
+	height := b.deployAt
+	if l1Height > b.bufferWindow {
+		height = l1Height - b.bufferWindow
 	}
-
-	if len(b.updates) == 0 || b.updates[0].L1Height > targetL1Height {
-		return b.initAddresses
+	if height > b.lastProcessedL1Height {
+		return false, fmt.Errorf("batcher address monitor is lagging behind, height: %d", l1Height)
 	}
 
-	// If the target L1 height is within the latest cached window, return the cached result.
-	// In a practical scenario, this is the most common case. Here means that during the time
-	// from `lastEventL1Height` to `l1Height`, the `events` are not changed. It is not needed to
-	// calculate valid batcher addresses.
-	latestCachedWindow := targetL1Height >= b.lastEventL1Height && targetL1Height <= b.lastProcessedL1Height
-	if b.cached && latestCachedWindow {
-		return b.cachedAddresses
-	}
-
-	result := map[common.Address]bool{}
-	for _, addr := range b.initAddresses {
-		result[addr] = true
-	}
-
-	for _, event := range b.updates {
-		if event.L1Height > targetL1Height {
+	index := 0
+	for i, updateHeight := range b.eventUpdatesAt {
```

Comment:
I think we should reverse this logic, basically we should accept a list of batcher addresses in the config. Then in case we see a new address which has not been a part of that list, we call the contract and see if that batcher address is valid or not and add it to the cache. Then the role of the batcher addr monitor should be to monitor events where a given "Batcher" is being removed from the list of trusted batcher. But for that it doesnt need to backfill, it can just start from the latest block and keep checking. It can trust that the addresses in the config are valid till the recent latest block and then going forward it auto updates its cache based on the address.